deep_qa/layers/tuple_matchers/slot_similarity_tuple_matcher.py

```diff
@@ -0,0 +1,146 @@
+from keras import backend as K
+from keras.layers import Layer
+from keras import initializations, activations
+from overrides import overrides
+
+from ...tensors.backend import apply_feed_forward
+from ...tensors.similarity_functions import similarity_functions
+
+
+class SlotSimilarityTupleMatcher(Layer):
+    """
+    Like other ``TupleMatch`` layers, this layer takes as input two tensors corresponding to two tuples,
+    an answer tuple and a background tuple, and calculates the degree to which the background tuple
+    `entails` the answer tuple.  In this layer, each input slot is represented as a dense embedding, so
+    to determine entailment we find the cosine similarity between these embedded slot representations,
+    i.e., the similarity between the first slot in each, then the second slot in each, etc.
+    This generates a set of similarity features equal to the number of slots in the tuples, which are
+    then fed to a shallow NN with output of size one.  The output of this NN is considered to be the
+    entailment score for the two tuples.
+
+    Inputs:
+        - tuple_1_input (the answer tuple), shape ``(batch size, num_slots, embed_dimensions)``.  There also
+          needs to be a corresponding mask of shape (batch size, num_slots) (or None) that indicates whether
+          a given slot was all padding.
+
+        - tuple_2_input (the background_tuple), shape ``(batch size, num_slots, embed_dimensions)``,
+          and again, there needs to be a corresponding mask of shape (batch size, num_slots) (or None)
+          that indicates whether a given slot was all padding.
+
+    Output:
+        - entailment score, shape ``(batch, 1)``
+
+    Parameters
+    ----------
+    - similarity_function_choice: str
```

Comment:
oh -- shoot, you meant that I need to pass in a dict for the *similarity_function*
ok -- let me work on that